graphrag/index/workflows/v1/subflows/create_final_text_units_pre_embedding.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2024 Microsoft Corporation.
+# Licensed under the MIT License
+
+"""All the steps to transform before we embed the text units."""
+
+from typing import cast
+
+import pandas as pd
+from datashaper.engine.verbs.verb_input import VerbInput
+from datashaper.engine.verbs.verbs_mapping import verb
+from datashaper.table_store.types import Table, VerbResult, create_verb_result
+from pandas._typing import Suffixes
+
+from graphrag.index.verbs.overrides.aggregate import aggregate_df
+
+
+@verb(
+    name="create_final_text_units_pre_embedding", treats_input_tables_as_immutable=True
+)
+def create_final_text_units_pre_embedding(
+    input: VerbInput,
+    covariates_enabled: bool = False,
+    **_kwargs: dict,
+) -> VerbResult:
+    """All the steps to transform before we embed the text units."""
+    table = input.get_input()
+    others = input.get_others()
+
+    selected = cast(Table, table[["id", "chunk", "document_ids", "n_tokens"]]).rename(
+        columns={"chunk": "text"}
+    )
+
+    final_entities = others[0]
+    final_relationships = others[1]
+    entity_join = _entities(final_entities)
+    relationship_join = _relationships(final_relationships)
+
+    entity_joined = _join(selected, entity_join)
+    relationship_joined = _join(entity_joined, relationship_join)
+    final_joined = relationship_joined
+
+    if covariates_enabled:
+        final_covariates = others[2]
+        covariate_join = _covariates(final_covariates)
+        final_joined = _join(relationship_joined, covariate_join)
+
+    aggregated = aggregate_df(
```

Comment:
Not sure if in scope for this PR, but surely we can optimize a lot if we don't use the aggregate verb but a different implementation. And it can be multiple one liners :)

Reply:
I will ignore for now so I can just focus on the collapsing, then we can do a separate sweep to cleanup and optimize a bunch of the pandas ops